pkg/ingester/encoding.go

```diff
@@ -25,84 +25,145 @@ const (
 
 // WALRecord is a struct combining the series and samples record.
 type WALRecord struct {
-	UserID     string
-	Series     []tsdb_record.RefSeries
-	RefEntries RefEntries
+	UserID string
+	Series []record.RefSeries
+
+	// entryIndexMap coordinates the RefEntries index associated with a particular fingerprint.
+	// This is helpful for constant time lookups during ingestion and is ignored when restoring
+	// from the WAL.
+	entryIndexMap map[uint64]int
+	RefEntries    []RefEntries
+}
+
+func (r *WALRecord) IsEmpty() bool {
+	return len(r.Series) == 0 && len(r.RefEntries) == 0
+}
+
+func (r *WALRecord) Reset() {
+	r.UserID = ""
+	if len(r.Series) > 0 {
+		r.Series = r.Series[:0]
+	}
+
+	for _, ref := range r.RefEntries {
+		recordPool.PutEntries(ref.Entries)
```

Comment:
nit or for later. This should be a leaky bucket. like https://github.com/grafana/loki/blob/6500f821951b737c980df832d9d78a34220430d6/pkg/chunkenc/pool.go#L49